src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNITcpHandle.cs

```diff
@@ -199,33 +199,78 @@ public SNITCPHandle(string serverName, int port, long timerExpire, object callba
                 throw new ArgumentOutOfRangeException(nameof(serverAddresses));
             }
 
-            // Try each address in turn, and return the socket opened for the first one that works.
-            ExceptionDispatchInfo lastException = null;
+            // Connect to each IP address in parallel, and return the socket opened for the first one that works.
+            List<Socket> sockets = new List<Socket>(serverAddresses.Length);
+            List<Task> connectTasks = new List<Task>(serverAddresses.Length);
+
+            object socketOpenTaskLock = new object();
+            Socket resultSocket = null;
+            ExceptionDispatchInfo lastSocketConnectEx = null;
+
+            CancellationTokenSource cancelTokenSource = new CancellationTokenSource();
+            CancellationToken cancelToken = cancelTokenSource.Token;
+
             foreach (IPAddress address in serverAddresses)
             {
-                var socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
+                Socket socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
                 try
                 {
-                    await socket.ConnectAsync(address, port).ConfigureAwait(false);
-                    return socket;
+                    sockets.Add(socket);
+                    Task socketTask = socket.ConnectAsync(address, port).ContinueWith(
+                        t =>
+                        {
+                            cancelToken.ThrowIfCancellationRequested();
+                            if (socket.Connected && !cancelToken.IsCancellationRequested)
+                            {
+                                lock (socketOpenTaskLock)
+                                {
+                                    if (resultSocket == null)
+                                    {
+                                        // Successful socket connection, so cancel other pending tasks
+                                        cancelTokenSource.Cancel();
+                                        resultSocket = socket;
```

Comment:
ConnectAsync's Task doesn't have a result. You just use the original socket after waiting for the connection to complete.